machine-runner/src/runner/runner-utils.ts

```diff
@@ -37,6 +37,7 @@ export type MachineEmitterEventMap<SwarmProtocolName extends string, MachineName
     nextState: unknown
   }) => unknown
   change: (_: StateOpaque<SwarmProtocolName, MachineName>) => unknown
+  caughtUpAndChanged: (_: StateOpaque<SwarmProtocolName, MachineName>) => unknown
```

Comment:
`change` which I assume is used when machine runner is used as an observable, retains its old behaviors basically so that it is usable in a reactive scenario e.g.:
- after a command is issued, but before the events are published, "change" is emitted. `machine.get()` will then return a nullified `state.commands`.
- on subscription error. `machine.get()` would then return null again
- on every caught up.

Reply:
Wouldn’t it be confusing to have an observable that yields different results that the `for await` loop? Users can still use the audit events to get these triggers if needed.

Reply:
It would be confusing if the user know beforehand that the `for await` loop trigger is the same as `on('change')`'s one.
Since the application code for the `for await` loop scenario and the `on('change')` scenario look totally different, it won't be confusing.

I think it is the other way around that if we want the user to subscribe to several event emitters, it would be a bit rigorous, especially since these separate event 'caughtUpAfterChange',  `command-rejection`, `command-issuance`, all of those are the internal mechanisms which are not visible to the eyes of the user.

Reply:
Hmm. These emitter event names are part of our API, so we need to make sure that they convey their meaning as accurately as possible. `change` is quite unspecific, and it currently also doesn’t emit on every conceivable change (i.e. it doesn’t emit when the `state.commands` turns `undefined` due to a command having been invoked.

`caughtUpAndChanged` include implementation details that users should probably not need to be aware of.

---

In summary:
- we need to determine which events are actually needed right now — we shouldn’t add events just because someone might use them in the future
- each event we do emit needs to have a name that gives the user the right intuition of when it is emitted

Reply:
`change` is more of a culture in the frontend world in a sense that `onChange` hook will be called if there is any meaningful change in the component that triggers the hook.

And we do have a `change` emission on `runner.ts` line 287. That is fired right after a `command` is called. This might be what makes it different between awaiting and not awaiting commands

Reply:
> each event we do emit needs to have a name that gives the user the right intuition of when it is emitted

What about `next`? It is unintuitive but we can make a promise that the AsyncIterable will be triggered on `next`, and then present the individual cases of when "next" is fired in the user manual.

Reply:
> does putting it on the emitter method overload work

I don't quite understand what is emitter method overload. Could you elaborate?

Reply:
The `.on('x', ...)` methods all have the same name but different signatures, distinguished by their first argument. I think I’ve seen them separately documented.

Reply:
> does putting it on the emitter method overload work?
If this means, `next` and `change` behave differently, then it does.